### DISCUSSION TITLE
About PNG to PSD Conversion

### DISCUSSION THREAD

**Discussion by ta-ota-hayato** [ImageMagick/ImageMagick#6146] [Help] [answered]
### ImageMagick version

7.1.0

### Operating system, version and so on

windows10

### Description

I would like to add a PNG file to each layer of a PSD file, is this possible with ImageMagick?

The following command will merge them into a single layer.
`magick.exe png_01.png png_02.png output_psd.psd`

Please help me. Thank you very much in advance.

**Comment by fmw42** [comment, marked as answer]
What do you mean by "add to each layer"?   Does that mean you already have existing layer images and want to merge in some other images? Or do you mean you want to create a PSD file from scratch using those images?

If you want to do the latter, then you have to recreated the first flattened layer from all your input images, thus

`magick.exe png_01.png png_02.png ( -clone 0,1 -flatten ) -insert 0 output_psd.psd`

**Reply by ta-ota-hayato** [reply]
Thank you for your reply.
What we would like to achieve is the following process
1. create a new psd
2. create layer_01 in psd and paste png_01.png
3. create layer_02 in psd and paste png_02.png
4. save the psd file

I tried the commands you gave me and achieved the results I was looking for.
However, if the sizes of png_01.png and png_02.png were different, the canvas sizes were not aligned.
Why is this?

Also, I would like to understand the following processing details.
Please let me know if you have any articles that would help me understand.
`( -clone 0,1 -flatten ) -insert 0`

**Reply by fmw42** [reply]
How do you want the PNGs pasted?  How would one know where they go if they are not the same size. You can always  use either -border, or -extent to pad them out. Or you can create a background color image and composite them in the location you want.

The ( -clone 0,1 -flatten ) -insert 0 takes all your input images (only two in this case) and flattens them and puts it as the first images (the -insert 0) to show the same thing you would see in Photoshop. This is the same as what you see in Photoshop when all layers are selected.

**Reply by ta-ota-hayato** [reply]
If you use png_01.png (w500 x h500), after creating a new layer (w500 x h500), you should ideally align the width and height and paste the png.

You are creating a layer to match the canvas size of the first image.
I will do a search on -border or -extent and see if what you are looking for is possible.
Thank you very much for your help in letting me know. It was very helpful.

**Comment by fmw42** [comment]
Imagemagick does not create layers first and paste the image. It just takes the images and puts them into PSD layers of the same size. So you have to pad your input images with some background, perhaps transparency or some color to match the size of the layers you want in the PSD file.

**Reply by ta-ota-hayato** [reply]
Okay, thank you very much.
Sorry, this is not the title of the article, but I was wondering about one last thing.
The layer names are "L1" and "L2", but is it possible to change them to "png_01", "png_02", etc.?

**Comment by fmw42** [comment]
As far as I know currently you cannot do that with Imagemagick.  An IM developer might comment further.

**Reply by ta-ota-hayato** [reply]
Understood.
If necessary, I will create another discussion and ask the question.
Thank you for your cooperation so far.

**Reply by dlemstra** [reply]
The -label is used as the name of the layer.

**Comment by snibgo** [comment]
> The layer names are "L1" and "L2", but is it possible to change them to "png_01", "png_02", etc.?

Looking at the code `coders/psd.c`, it seems that IM will use the image `label` property if it is set. Otherwise, it uses `L1` or `L2` etc.

But I don't use PSD files or Photoshop.

**Reply by ta-ota-hayato** [reply]
By adding `-label "%t" png_01.png png_02.png` I was able to change the layer name.
Thank you both for letting me know.

Reference: https://imagemagick.org/script/escape.php
